### ISSUE TITLE
Indexing is not idempotent using Base.IdentityUnitRange

### ISSUE THREAD

**Issue by jishnub** [JuliaArrays/OffsetArrays.jl#155]
This does not satisfy `a[ax][i] == a[ax[i]]`

```julia
julia> a = OffsetVector(3:4, 2:3)
3:4 with indices 2:3

julia> ax = Base.IdentityUnitRange(2:3)
Base.IdentityUnitRange(2:3)

julia> a[ax]
3:4

julia> a[ax][2]
4

julia> a[ax[2]]
3
```

The issue seems to be that `a[ax]` results in an `UnitRange` instead of an `OffsetVector`, and has different axes from `ax`. Eg something like this would be worse:

```julia
julia> a = OffsetVector(3:4, 10:11)
3:4 with indices 10:11

julia> ax = Base.IdentityUnitRange(10:11)
Base.IdentityUnitRange(10:11)

julia> a[ax]
3:4

julia> a[ax[10]]
3

julia> a[ax][10]
ERROR: BoundsError: attempt to access 2-element UnitRange{Int64} at index [10]
Stacktrace:
 [1] throw_boundserror(::UnitRange{Int64}, ::Int64) at ./abstractarray.jl:541
 [2] getindex(::UnitRange{Int64}, ::Int64) at ./range.jl:646
 [3] top-level scope at REPL[26]:1
```

**Comment by jishnub** [comment]
Looks like this happens because of https://github.com/JuliaArrays/OffsetArrays.jl/blob/dc867d5fc8b9e135b34ec60331c7ff781bcb1a1c/src/OffsetArrays.jl#L309

and the fact that `Base.IdentityUnitRange` gets transformed to an `UnitRange` upon broadcasting

**Comment by johnnychen94** [comment]
I'm rewriting the constructor tests in #148. I'll add one test for it!

**Comment by jishnub** [comment]
Are you fixing this? I wonder what's the best approach? One way to solve this would be 

```julia
Base.getindex(a::OffsetRange, r::AbstractRange) = OffsetArray(a.parent[r .- a.offsets[1]], axes(r,1))
```

however this causes tests to fail, as they explicitly test for `UnitRange`s. A temporary fix might be to add an extra method for `Base.IdentityUnitRange`, but this isn't ideal.

**Comment by johnnychen94** [comment]
Not planned in that PR. I'll tag it as broken.

**Comment by timholy** [comment]
Ranges are a mess and this is what https://github.com/JuliaLang/julia/pull/30950 is about. I think what happened is that initially we intended to keep `AbstractRange` with 1-indexing but then we realized rather late in the process that it was really useful to break that requirement, and the process never got finished properly. Now we have to figure out how to fix it in a non-breaking way, which is what made that PR scary and seem like it might be heading down the wrong path.
